File: server/src/modules/exportService.ts
import type { OTSlot, ShiftRoster, AllFillRates } from '../types.js';
import { getWeek } from './fillRateCalculator.js';

function escapeCSVField(field: string): string {
  const str = String(field);
  if (str.includes(',') || str.includes('"') || str.includes('\n')) {
    return `"${str.replace(/"/g, '""')}"`;
  }
  return str;
}

/**
 * Serializes slots to CSV string.
 */
export function serializeSlotsToCSV(slots: OTSlot[], roster: ShiftRoster): string {
  // Build agent → manager mapping from roster
  const agentToManager = new Map<string, string>();
  const programToManager = new Map<string, string>();
  for (const entry of roster.entries) {
    if (!agentToManager.has(entry.agent)) {
      agentToManager.set(entry.agent, entry.manager);
    }
    if (!programToManager.has(entry.program)) {
      programToManager.set(entry.program, entry.manager);
    }
  }

  const headers = ['Date', 'Week', 'Program', 'OT_Type', 'Time_Window', 'Slot_Status', 'Assigned_Agent', 'Manager'];

  const rows = slots.map((slot) => {
    const week = getWeek(slot.date);
    const assignedAgent = slot.filledByAgentName ?? slot.assignedAgentName ?? '';
    let manager = '';
    if (slot.filledByAgentId) {
      manager = agentToManager.get(slot.filledByAgentId) ?? '';
    }
    if (!manager && slot.assignedAgentId) {
      manager = agentToManager.get(slot.assignedAgentId) ?? '';
    }
    if (!manager) {
      manager = programToManager.get(slot.program) ?? '';
    }

    return [slot.date, week, slot.program, slot.otType, slot.timeWindow, slot.status, assignedAgent, manager]
      .map(escapeCSVField)
      .join(',');
  });

  return [headers.join(','), ...rows].join('\n');
}

/**
 * Serializes fill rates to CSV string.
 */
export function serializeFillRatesToCSV(fillRates: AllFillRates): string {
  const headers = ['Grouping', 'Key', 'Total_Released', 'Total_Filled', 'Fill_Rate'];
  const rows: string[] = [];

  // Overall
  const o = fillRates.overall;
  rows.push(['Overall', 'All', String(o.totalReleased), String(o.totalFilled), o.fillRate !== null ? String(o.fillRate) : 'N/A'].map(escapeCSVField).join(','));

  // By Program
  for (const [key, val] of Object.entries(fillRates.byProgram)) {
    rows.push(['Program', key, String(val.totalReleased), String(val.totalFilled), val.fillRate !== null ? String(val.fillRate) : 'N/A'].map(escapeCSVField).join(','));
  }

  // By Manager
  for (const [key, val] of Object.entries(fillRates.byManager)) {
    rows.push(['Manager', key, String(val.totalReleased), String(val.totalFilled), val.fillRate !== null ? String(val.fillRate) : 'N/A'].map(escapeCSVField).join(','));
  }

  // By Date
  for (const [key, val] of Object.entries(fillRates.byDate)) {
    rows.push(['Date', key, String(val.totalReleased), String(val.totalFilled), val.fillRate !== null ? String(val.fillRate) : 'N/A'].map(escapeCSVField).join(','));
  }

  // By Week
  for (const [key, val] of Object.entries(fillRates.byWeek)) {
    rows.push(['Week', key, String(val.totalReleased), String(val.totalFilled), val.fillRate !== null ? String(val.fillRate) : 'N/A'].map(escapeCSVField).join(','));
  }

  return [headers.join(','), ...rows].join('\n');
}

File: python-server/modules/export_service.py
"""
CSV export service.
Mirrors exportService.ts exactly.
"""
from .fill_rate_calculator import get_week


def _escape_csv_field(field):
    s = str(field)
    if ',' in s or '"' in s or '\n' in s:
        return '"' + s.replace('"', '""') + '"'
    return s


def serialize_slots_to_csv(slots, roster):
    """Serializes slots to CSV string."""
    agent_to_manager = {}
    program_to_manager = {}
    for entry in roster.get('entries', []):
        if entry['agent'] not in agent_to_manager:
            agent_to_manager[entry['agent']] = entry['manager']
        if entry['program'] not in program_to_manager:
            program_to_manager[entry['program']] = entry['manager']

    headers = ['Date', 'Week', 'Program', 'OT_Type', 'Time_Window', 'Slot_Status', 'Assigned_Agent', 'Manager']

    rows = []
    for slot in slots:
        week = get_week(slot['date'])
        assigned_agent = slot.get('filledByAgentName') or slot.get('assignedAgentName') or ''
        manager = ''
        if slot.get('filledByAgentId'):
            manager = agent_to_manager.get(slot['filledByAgentId'], '')
        if not manager and slot.get('assignedAgentId'):
            manager = agent_to_manager.get(slot['assignedAgentId'], '')
        if not manager:
            manager = program_to_manager.get(slot['program'], '')

        row_data = [slot['date'], week, slot['program'], slot['otType'], slot['timeWindow'], slot['status'], assigned_agent, manager]
        rows.append(','.join(_escape_csv_field(f) for f in row_data))

    return '\n'.join([','.join(headers)] + rows)


def serialize_fill_rates_to_csv(fill_rates):
    """Serializes fill rates to CSV string."""
    headers = ['Grouping', 'Key', 'Total_Released', 'Total_Filled', 'Fill_Rate']
    rows = []

    # Overall
    o = fill_rates['overall']
    fr = str(o['fillRate']) if o['fillRate'] is not None else 'N/A'
    rows.append(','.join(_escape_csv_field(f) for f in ['Overall', 'All', str(o['totalReleased']), str(o['totalFilled']), fr]))

    # By Program
    for key, val in fill_rates.get('byProgram', {}).items():
        fr = str(val['fillRate']) if val['fillRate'] is not None else 'N/A'
        rows.append(','.join(_escape_csv_field(f) for f in ['Program', key, str(val['totalReleased']), str(val['totalFilled']), fr]))

    # By Manager
    for key, val in fill_rates.get('byManager', {}).items():
        fr = str(val['fillRate']) if val['fillRate'] is not None else 'N/A'
        rows.append(','.join(_escape_csv_field(f) for f in ['Manager', key, str(val['totalReleased']), str(val['totalFilled']), fr]))

    # By Date
    for key, val in fill_rates.get('byDate', {}).items():
        fr = str(val['fillRate']) if val['fillRate'] is not None else 'N/A'
        rows.append(','.join(_escape_csv_field(f) for f in ['Date', key, str(val['totalReleased']), str(val['totalFilled']), fr]))

    # By Week
    for key, val in fill_rates.get('byWeek', {}).items():
        fr = str(val['fillRate']) if val['fillRate'] is not None else 'N/A'
        rows.append(','.join(_escape_csv_field(f) for f in ['Week', key, str(val['totalReleased']), str(val['totalFilled']), fr]))

    return '\n'.join([','.join(headers)] + rows)
